Оптимизируй это (часть 2)

VM-Study opened the pull request:
---
:mortar_board: [Оптимизируй это (часть 2)](https://up.htmlacademy.ru/react/14/user/2342349/tasks/15)

Commit:
8.13 (part 2)

Commit:
8.13 (part 2) Refactor code, create "LoadingText" component

potyoma commented on src/components/favorite-list/favorite-list.tsx:
```diff
@@ -7,38 +9,43 @@ interface FavoriteListProps {
       return acc;
     }, {});
 
+  const favoriteList = Object.entries(cityOffersMap).map(([cityName, cityOffers]) => (
```
отдельный компонент

potyoma commented on src/components/loading-text/loading-text.tsx:
```diff
@@ -0,0 +1,9 @@
+import styles from './loading-text.module.css';
+
+function loadingText() {
```
с заглавной

potyoma commented on src/components/sort-list/sort-list.tsx:
```diff
@@ -26,6 +26,18 @@ function SortList() {
     }
   }
 
+  const SortOptionList = Object.entries(SortOptions).map(([key , {title}], index) => (
```
вместе с ul вынести в отдельный компонент

potyoma commented on src/pages/main-page/main-page.tsx:
```diff
@@ -1,18 +1,42 @@
   const dispatch = useAppDispatch();
+  const isOffersEmpty = offers.length === 0;
+
+  const offerList = isOffersEmpty ? (
```
отдельным компонентом

potyoma commented on src/pages/offer-page/offer-page.tsx:
```diff
@@ -61,6 +68,53 @@ function OfferPage() {
       getMapDataFromOffers([currentOffer, ...currentNearbyOffers.slice(0, MAX_NEAR_OFFERS)], selectedCity, selectedOfferId);
   }
 
+  const mapBlock = (mapCity && mapPoints) ? (
```
через hoc, который будет принимать условие рендера и выводить компонент на его основе

potyoma commented on src/pages/offer-page/offer-page.tsx:
```diff
@@ -61,6 +68,53 @@ function OfferPage() {
+    <ReviewList reviews={currentReviews.slice(0, MAX_REVIEWS_PER_OFFER)}/>
+  ) : <LoadingText/>;
+
+  const offerBlock = currentOffer ? (
```
компонент

potyoma commented on src/store/api-communication/api-communication.slice.ts:
```diff
@@ -128,6 +132,38 @@ export const apiCommunicationSlice = createSlice({
+      .addCase(fetchFavoritesAction.rejected, (state, action) => {
+        state.isLoading = false;
+        if (typeof action.payload === 'string') {
+          toast.warning(action.payload);
```
перенести сайд эффект из редьюсера

potyoma commented on src/components/navigation-menu/navigation-menu.tsx:
```diff
@@ -4,12 +4,16 @@ import {AppRoute, AuthorizationStatus, AuthorizationStatusType} from '../../cons
 }
 
-function NavigationMenu({authorizationStatus}: Readonly<NavigationMenuProps>) {
+function NavigationMenu({authorizationStatus, userLogin, userAvatarUrl, favorites}: Readonly<NavigationMenuProps>) {
```
вынести получение данных пользователя в компоненте в отдельный хук

Commit:
Add FavoriteCityList, OfferBlock, OfferListContainer, SortOptionList components.
Add useUserData hook. Modify api-communication slice, user-api-actions, and data-api-actions to handle toasts for successful and unsuccessful actions.

potyoma commented on src/components/offer-block/offer-block.tsx:
```diff
@@ -0,0 +1,94 @@
+  const currentReviews = useAppSelector(getCurrentReviews);
+  const authorizationStatus = useAppSelector(getAuthorizationStatus);
+  const selectedCity = useAppSelector(getSelectedCity);
+  const MemoizedOfferDetails = memo(OfferDetails);
```
вне тела компонента

potyoma commented on src/components/offer-block/offer-block.tsx:
```diff
@@ -0,0 +1,94 @@
+      getMapDataFromOffers([currentOffer, ...currentNearbyOffers.slice(0, MAX_NEAR_OFFERS)], selectedCity, selectedOfferId);
+  }
+
+  const mapBlock = (mapCity && mapPoints) ? (
```
вынеси логику в функцию. повторяется

potyoma commented on src/components/offer-block/offer-block.tsx:
```diff
@@ -0,0 +1,94 @@
+    <MemoizedReviewList reviews={currentReviews.slice(0, MAX_REVIEWS_PER_OFFER)}/>
+  ) : <LoadingText/>;
+
+  return currentOffer ? (
```
смотри выше

potyoma commented on src/components/sort-option-list/sort-option-list.tsx:
```diff
@@ -0,0 +1,47 @@
+    toggleSortMenu();
+  }
+
+  function handleKeyDown(event: React.KeyboardEvent, action: () => void) {
```
handleKeyDown(action) {
    return (event) => {
      ...
    }
}

potyoma replied:
каррирование (currying)

potyoma commented on src/components/sort-option-list/sort-option-list.tsx:
```diff
@@ -0,0 +1,47 @@
+      className={classNames('places__option', {'places__option--active': currentSortType === key})}
+      tabIndex={index}
+      onClick={() => handleSortChange(key as keyof typeof SortOptions)}
+      onKeyDown={(event) => handleKeyDown(event, () => handleSortChange(key as keyof typeof SortOptions))}
```
handleKeyDown(() => handleSprtChange)
